src/streams.rs

```diff
@@ -0,0 +1,698 @@
+                    })
+                    .collect::<Vec<StreamKey>>()
+            })
+            .collect();
+        Ok(StreamReadReply { keys })
+    }
+}
+
+impl FromRedisValue for StreamRangeReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: Vec<HashMap<String, HashMap<String, Value>>> = from_redis_value(v)?;
+        let ids: Vec<StreamId> = rows
+            .into_iter()
+            .flat_map(|row| {
+                row.into_iter()
+                    .map(|(id, map)| StreamId { id, map })
+                    .collect::<Vec<StreamId>>()
+            })
+            .collect();
+        Ok(StreamRangeReply { ids })
+    }
+}
+
+impl FromRedisValue for StreamClaimReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: Vec<HashMap<String, HashMap<String, Value>>> = from_redis_value(v)?;
+        let ids: Vec<StreamId> = rows
+            .into_iter()
+            .flat_map(|row| {
+                row.into_iter()
+                    .map(|(id, map)| StreamId { id, map })
+                    .collect::<Vec<StreamId>>()
+            })
+            .collect();
+        Ok(StreamClaimReply { ids })
+    }
+}
+
+type SPRInner = (
+    usize,
+    Option<String>,
+    Option<String>,
+    Vec<Option<(String, String)>>,
+);
+impl FromRedisValue for StreamPendingReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let (count, start, end, consumer_data): SPRInner = from_redis_value(v)?;
+
+        if count == 0 {
+            Ok(StreamPendingReply::Empty)
+        } else {
+            let mut result = StreamPendingData::default();
+
+            let start_id = start.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::Other,
+                    "IllegalState: Non-zero pending expects start id",
+                )
+            })?;
+
+            let end_id = end.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::Other,
+                    "IllegalState: Non-zero pending expects end id",
+                )
+            })?;
+
+            result.count = count;
+            result.start_id = start_id;
+            result.end_id = end_id;
+
+            for cd in consumer_data {
+                if let Some((name, pending)) = cd {
+                    let mut info = StreamInfoConsumer::default();
+                    info.name = name;
+                    if let Ok(v) = pending.parse::<usize>() {
+                        info.pending = v;
+                    }
+                    result.consumers.push(info);
+                }
+            }
+
+            Ok(StreamPendingReply::Data(result))
+        }
+    }
+}
+
+impl FromRedisValue for StreamPendingCountReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let mut reply = StreamPendingCountReply::default();
+        match v {
+            Value::Bulk(outer_tuple) => {
+                for outer in outer_tuple {
```

Comment:
Willing to rewrite this as a `flat_map` if that's your preference.  The logic is, at least, working, in this version. :microscope:

Reply:
I'm not sure it would make things any better to take a more functional approach here.  It would appear there is the potential for nested errors that then need to be traversed in order to figure out that the function failed.

Here's a draft of that approach:

```rust
    if let Value::Bulk(outer_tuple) = v {
            let fm: Vec<Result<bool, Error>> = outer_tuple
                .into_iter()
                .map(|outer| {
                    if let Value::Bulk(inner_tuple) = outer {
                        Ok(true)
                    } else {
                        Err(todo!())
                    }
                })
                .collect();
```